1580 document web modeler public api (#1581)
camunda/camunda-platform-docs: docs/apis-clients/web-modeler-api/index.md

Requests and responses are in JSON notation."
---

:::caution Beta Offering
Web Modeler API is currently offered as a [beta release](../../reference/early-access#beta). It is not recommended for production use and there is no maintenance service guaranteed.

While in beta, the API may introduce breaking changes without prior notice.

However, we encourage you to provide feedback via your designated support channel or the [Camunda Forum](https://forum.camunda.io/).
:::

Web Modeler provides a REST API at `/api/*`. Clients can access this API by passing a JWT access token in an authorization header `Authorization: Bearer <JWT>`.

## OpenAPI documentation

A detailed API description is available as [OpenAPI](https://www.openapis.org/) specification at [https://modeler.cloud.camunda.io/swagger-ui/index.html](https://modeler.cloud.camunda.io/swagger-ui/index.html).

## Authentication

To authenticate for the API, generate a JWT token and pass it in each request; guidance on this is provided in the following sections.

### Obtain JWT token

**Example:**

1. Create client credentials by clicking **Console > Manage (Organization) > Console API > Create New Credentials**.

> 2. Add permissions to this client for **Web Modeler API (beta)**.

Reviewer: Are there specific steps to accomplish this?

Author: Yes, but trivial ones. Once you create credentials you will see a UI with check boxes for Web Modeler. 
![image](https://user-images.githubusercontent.com/14032870/209138743-143d6490-6afc-4bc2-a93c-06bae9c761ab.png)

Screenshot might be helpful, but it is pretty much self explanatory, in my mind. Also since this is the first iteration of documentation on a beta feature, I think it can be a little rough around the edges.

Personally, I would prefer a page like this https://docs.camunda.io/docs/components/console/manage-clusters/manage-api-clients/ for Console API clients.

The page I linked is about managing API clients on a per-cluster basis. There is also a way to manage clients/credentials on an organization-wide basis. For this, there is currently no documentation. If there would be, I would link to it.

> 3. After creating the client, you can download a shell script to obtain a token.

Reviewer: How do you download this?

Author: You click on the download button that appears after you clicked "Create" in the last UI. I admit that it is not clear if you only read this text, but if you read this text and do it at the same time, it should be pretty obvious. Also same comment applies to missing documentation on creating Console API credentials as above:

![image](https://user-images.githubusercontent.com/14032870/209139863-811fa40c-04da-499b-b828-1b96be2f7692.png)

> 4. When you run it, you will get something like the following:

Reviewer: When you run the shell script?

   ```json
   {
     "access_token": "eyJhbG...",
     "expires_in": 300,
     "refresh_expires_in": 0,
     "token_type": "Bearer",
     "not-before-policy": 0
   }
   ```

### Use JWT token

1. Take the **access_token** value from the response object and store it as your token.
2. Send the token as an authorization header in each request. In this case, call the info endpoint to validate the token.
   ```shell
   curl -X POST 'http://modeler.cloud.camunda.io/api/beta/info' -H 'Content-Type: application/json' -H 'Authorization: Bearer eyJhb...' -d '{}'
   ```
3. You will get something like the following:
   ```json
   {
     "version": "beta",
     "authorizedOrganization": "12345678-ABCD-DCBA-ABCD-123456789ABC",
     "createPermission": true,
     "readPermission": true,
     "updatePermission": true,
     "deletePermission": false
   }
   ```

> ## Good to know

Reviewer: We can likely remove this section altogether since we currently only have one topic. If this expands, I would suggest naming it something like "FAQ" instead. Therefore, ### Names versus ids can become a H2.

Most of the API is self-explanatory. This section describes details which are less obvious.

### Names versus ids

> In Web Modeler you can have multiple files with the same name, multiple folders with the same name, and even multiple projects with the same name. Internally, duplicate names are disambiguated by unique ids.

Reviewer: What do you mean by disambiguated here? I might use a simpler word.

---
id: index
title: Web Modeler API (beta) (REST)
description: "Web Modeler Public API (beta) is a REST API and provides access to Web Modeler Data.
Requests and responses are in JSON notation."
---

:::caution Beta Offering
Web Modeler API is currently offered as a [beta release](../../reference/early-access#beta). It is not recommended for production use and there is no maintenance service guaranteed.

> Special [terms & conditions](https://camunda.com/legal/terms/camunda-platform/camunda-platform-8-self-managed/) apply.

Author: good catch

While in beta the API may introduce breaking changes without prior notice.

However, we encourage you to provide feedback via your designated support channel or the [Camunda Forum](https://forum.camunda.io/).
:::

## OpenAPI documentation

A detailed API description is also available as [OpenAPI](https://www.openapis.org/) specification at [https://modeler.cloud.camunda.io/swagger-ui/index.html](https://modeler.cloud.camunda.io/swagger-ui/index.html)

## Authentication

To authenticate for the API you need to generate a JWT token and pass it in each request.

### Obtain JWT token

Web Modeler provides a REST API under the endpoint `/api`. Clients can access this API using a JWT access token in an authorization header `Authorization: Bearer <JWT>`.

**Example:**

> 1. [Create an API Client](../../../components/console/manage-clusters/manage-api-clients/#create-a-client).

Author: good catch

2. Add permissions to this client for **Web Modeler Public API (beta)**.

> 3. After creating the client, you can download a shell script to obtain a token

Reviewer 2: ```suggestion
3. After creating the client, you can download a shell script to obtain a token.
```

4. When you run it, you will get something like the following:
   ```json
   {
     "access_token": "eyJhbG...",

>      "expires_in": 300,

Author: No. I am not sure refresh tokens are a thing for public APIs. In UIs it's different, because you have an open website or a session. So you have something where you can register a timer and refresh the token.

With M2M clients this is less common. Usually, you assume stateless and keep the token for as long as it is valid. Getting a new token vs. refreshing a token also makes little difference, technically. So I would not describe this, unless there is a need for it.

     "refresh_expires_in": 0,
     "token_type": "Bearer",
     "not-before-policy": 0
   }
   ```

## Use JWT token

1. Take the **access_token** value from the response object and store it as your token.
2. Send the token as an authorization header in each request. In this case, call the info endpoint to validate the token.
   ```shell
   curl -X POST 'http://modeler.cloud.camunda.io/api/beta/info' -H 'Content-Type: application/json' -H 'Authorization: Bearer eyJhb...' -d '{}'
   ```
3. You will get something like the following:
   ```json
   {
     "version": "beta",
     "authorizedOrganization": "12345678-ABCD-DCBA-ABCD-123456789ABC",
     "createPermission": true,
     "readPermission": true,
     "updatePermission": true,
     "deletePermission": false
   }
   ```

## Good to know

Most of the API is self-explanatory. This section describes details which are less obvious.

### Names versus ids

In Web Modeler you can have multiple files with the same name, multiple folders with the same name and even multiple projects with the same name. Internally, duplicate names are disambiguated by unique ids.

> The API gives you access to the names as well as the Ids. For example when requesting a file you will find in it the following information:

Reviewer 2: ```suggestion
The API gives you access to the names as well as the Ids. For example, when requesting a file you will get the following information:
```

> - **simplePath** gives you access to the human-readable path. This path may be ambiguous or may have ambiguous elements (e.g. folders) in it

Reviewer 2: ```suggestion
- **simplePath** contains the human-readable path. This path may be ambiguous or may have ambiguous elements (e.g. folders) in it.
```

> - **canonicalPath** gives you access to the unique passed. It is a list of **PathElementDto** objects which contain the id and the name of the element

Reviewer 2: ```suggestion
- **canonicalPath** contains the unique path. It is a list of **PathElementDto** objects which contain the id and the name of the element.
```

> - **canonicalPath** gives you access to the unique passed. It is a list of **PathElementDto** objects which contain the id and the name of the element

Reviewer 2: Maybe it would make sense to add an example here?

Author: Maybe on the second iteration :clown_face:

> Internally, the ids are what matters. You can rename files, and move files between folders and projects and the id will stay the same.

Reviewer 2: I think "ID" (instead of lowercase "id") is the correct spelling, so I'd suggest to use "ID" and "IDs" everywhere.

Author: @christinaausley Do you agree?

Reviewer: So this has been an interesting conversation since I started working with Camunda. Currently in our [Technical Writing Glossary](https://confluence.camunda.com/display/HAN/Technical+Writing+Glossary) [we use id](https://confluence.camunda.com/display/HAN/Technical+Writing+Glossary#:~:text=Id,For%20example%2C%20%60regionId%60.).

This is to mirror API responses/JSON values exactly as they are referenced in the text. Do you have strong thoughts either way here @jfriedenstab?

Reviewer 2: I don't have strong thoughts here :). In my opinion, especially the plural "IDs" is easier to read than "ids", but maybe that's just my impression. If we use "id" in the technical writing glossary, then let's just use it here, too. I also get your point regarding the JSON response.

By the way, this was my reference: https://dictionary.cambridge.org/dictionary/english/id

> "id": in psychoanalysis, the deepest part of the unconscious mind

> "ID": a set of numbers, letters, or symbols that is used to represent a piece of data or a process in a computer program

Author: I would keep it as is then, until there is an update on the technical writing glossary
